src/caller.jl

```diff
@@ -0,0 +1,189 @@
+            return parent(i)
+        elseif current_ptr isa ideal
+            return parent(i).base_ring
+        end
+    end
+    return ring
+end
+
+function prepare_argument(x::Array{Int64,1})
+    return Any[ mapping_types_reversed[:INTVEC_CMD], libSingular.jl_array_to_intvec(x) ], nothing
+end
+
+function prepare_argument(x::Array{Int64,2})
+    return Any[ mapping_types_reversed[:INTMAT_CMD], libSingular.jl_array_to_intmat(x) ], nothing
+end
+
+function prepare_argument(x::Int64)
+    return Any[ mapping_types_reversed[:INT_CMD], Ptr{Cvoid}(x) ], nothing
+end
+
+function prepare_argument(x::String)
+    return Any[ mapping_types_reversed[:STRING_CMD], libSingular.copy_string_to_void(x) ], nothing
+end
+
+function prepare_argument(x::PolyRing)
+    new_ptr = libSingular.get_ring_ref(x.ptr)
+    return Any[ mapping_types_reversed[:RING_CMD], new_ptr ], x
+end
+
+function prepare_argument(x::spoly)
+    rng = parent(x)
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:POLY_CMD], libSingular.copy_polyptr_to_void(x.ptr, rng_ptr) ], rng
+end
+
+function prepare_argument(x::svector)
+    rng = parent(x).base_ring
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:VECTOR_CMD], libSingular.copy_polyptr_to_void(x.ptr, rng_ptr) ], rng
+end
+
+function prepare_argument(x::sideal)
+    rng = parent(x).base_ring
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:IDEAL_CMD], libSingular.copy_idealptr_to_void(x.ptr, rng_ptr)], rng
+end
+
+function prepare_argument(x::smodule)
+    rng = parent(x).base_ring
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:MODUL_CMD], libSingular.copy_idealptr_to_void(x.ptr, rng_ptr)], rng
+end
+
+function prepare_argument(x::sresolution)
+    rng = base_ring(x)
+    res = Any[ mapping_types_reversed[:RESOLUTION_CMD], libSingular.create_syStrategy_data(x.ptr, rng.ptr) ]
+    return res, rng
+end
+
+function prepare_argument(x::Any)
+    if x.ptr isa libSingular.number
+        ptr = x.ptr
+        rng = parent(x)
+        new_ptr = libSingular.n_Copy(ptr, rng.ptr)
+        return Any[ mapping_types_reversed[:NUMBER_CMD], new_ptr.cpp_object ], nothing
+    elseif x.ptr isa libSingular.ip_smatrix
+        rng = parent(x)
+        return Any[ mapping_types_reversed[:MATRIX_CMD], libSingular.mpCopy(x.ptr, rng.ptr).cpp_object ], rng
+    elseif x.ptr isa libSingular.__mpz_struct
+        return Any[ mapping_types_reversed[:BIGINT_CMD], x.ptr.cpp_object ], nothing
+    elseif x.ptr isa libSingular.sip_smap
+        return Any[ mapping_types_reversed[:MAP_CMD], x.ptr.cpp_object ], nothing
+    elseif x.ptr isa libSingular.bigintmat
+        return Any[ mapping_types_reversed[:BIGINTMAT_CMD], x.ptr.cpp_object ], nothing
+    end
+
+end
+
+function low_level_caller_rng(lib::String, name::String, ring, args...)
+    libSingular.load_library(lib)
+    arguments = [prepare_argument(i) for i in args]
+    arguments = Any[ i for (i, j) in arguments ]
```

Comment:
So you throw away the rings here. Is this safe? What will happen if I pass in an argument that lives in a different ring? At least in the past, this could crash Singular, IIRC (but perhaps I am wrong? @hannes14 will know better :-)

Reply:
Ideally, we would compare rings, yes.